### PR TITLE
Implement DistanceAboveTerrain

### PR DESCRIPTION
**Pull request by phrohdoh**
#7904 sans-aircraft.

cc @pchote, @penev92

Test by (un)loading cargo on elevated terrain in TS/RA2.

Fixes https://github.com/OpenRA/OpenRA/issues/8573.
Fixes https://github.com/OpenRA/OpenRA/issues/8780.

### CODE REVIEW

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

**Comment by pchote** [comment]
I don't really like the `TerrainHeightAt` abstraction because you're saying "at" but then completely ignoring the z coordinate.  I was working on this feature plus proper slope height yesterday (before I hit and was forced to fix #8463), and decided on `DistanceAboveTerrain` which uses everything in the given WPos, matches the `DistanceToEdge` function, and returns a result that is directly useful for `actor.IsAtGroundLevel` (check if == 0) and `VoxelRenderable` (just subtract it from the position).

**Comment by pchote** [comment]
Could you also please use `IsAtGroundLevel` at Crate.csL193 to fix #8780?

**Comment by GraionDilach** [comment]
It should be at Crushable imo, not just crates, but the whole crushing logic has that z check.

**Comment by Phrohdoh** [comment]
Updated to reflect the requested changes.

**Comment by reaperrr** [comment]
The `isAir` and `isWater` checks in `CreateEffectWarhead` need to be adapted as well, otherwise they will fail on elevated TS terrain.

File: OpenRA.Mods.Common/ActorExts.cs [hunk 1]
```diff
@@ -18,6 +18,28 @@ namespace OpenRA.Mods.Common
...
+
+			if (!self.IsInWorld)
+				return false;
+
+			var map = self.World.Map;
+
+			if (!map.Contains(self.Location))
+				return false;
+
+			var centerPos = self.CenterPosition;
```
Comment (pchote) [inline OpenRA.Mods.Common/ActorExts.cs]: You aren't doing anything interesting with these temporary variables, so just `return map.DistanceAboveTerrain(self.CenterPosition) == 0`.

**Comment by Phrohdoh** [comment]
Okay, fixed.
But do note:

>[12:16:44] Phrohdoh: okay, i can fix crates as long as you're not going to ask for a bunch of other fixes
[12:16:57] Phrohdoh: the point was to do as little as possible so it doesn't become the other pr
[12:17:12] pchote: those are my only requests
[12:17:16] Phrohdoh: okay

**Comment by GraionDilach** [comment]
Well, I have no further requests regarding this.

**Comment by pchote** [comment]
Confirmed that this fixes transports, but not crates.

**Comment by pchote** [comment]
There's z == 0 check at Mobile.cs:L568 that is breaking crates.  Can you please fix that too?

**Comment by Phrohdoh** [comment]
Thanks, done.

**Comment by pchote** [comment]
Confirmed fixes :+1:

**Pull request merged**